experimental/torch_xla2/torch_xla2/export.py

```diff
@@ -104,3 +107,130 @@ def func(states, inputs):
+    JAX symbolic to map to for a given torch tensor.
+    """
+    if len(range_constraints) == 0:
+      return None
+
+    def _build_symbolic_constraints(symbol_name, torch_constraint):
+      """Convert torch SymInt constraints to string for JAX symbolic_shape
+      Using sympy may be overkill here, currently PyTorch only uses ValueRanges
+      which allow specifying the min and the max of a value, for example:
+        torch.export.Dim("a", min=5, max=10)
+         ==> ("a >= 5", "a <= 10",)
+      """
+      if not isinstance(torch_constraint, torch.utils._sympy.value_ranges.ValueRanges) or torch_constraint.is_bool:
+        raise TypeError(f"No symbolic constraint handler for: {constraint}")
+
+      constraints = []
+      symbol = sympy.Symbol(symbol_name)
+      if (torch_constraint.lower != 2):
+        constraints.append(symbol >= torch_constraint.lower)
+      if (not torch_constraint.upper.is_infinite):
+        constraints.append(symbol <= torch_constraint.upper)
+      
+      return tuple(sympy.pretty(c, use_unicode=False) for c in constraints)
+
+    def _build_symbolic_shape(sym, constraint, free_symbols):
+      """Returns a JAX symbolic shape for a given symbol and constraint
+
+      There are two possible sympy `sym` inputs:
+        1. Symbol - (s0) These can have custom constraints.
+        2. Expr - (s0*2) These apply the expr to s0's constraints, cannot override.
+      
+        Currently support is limited to operations with a symbol and and int,
+        in `torch/export/dynamic_shapes.py`:
+        "Only increasing linear operations with integer coefficients are supported."
+      """
+      symbol_name = str(sym)
+      constraints = _build_symbolic_constraints(symbol_name, constraint)
+      if sym.is_symbol:
+        symbolic_shape = jax.experimental.export.symbolic_shape(symbol_name, constraints=constraints)
+      else:
+        assert len(sym.free_symbols) > 0
+        scope = free_symbols[str(list(sym.free_symbols)[0])].scope
+        symbolic_shape = jax.experimental.export.symbolic_shape(symbol_name, scope=scope)
+      assert len(symbolic_shape) == 1
+      return symbolic_shape[0]
+
+    # Populate symbol variables before expressions, exprs need to use the same
+    # Symbolic scope as the variable they operate on. Expressions can only be
+    # integer compuations on symbol variables, so each symbol variable is OK to
+    # have its own scope.
+    symbolic_shapes = {}
+    symbol_variables = [(s,v) for s,v in range_constraints.items() if s.is_symbol]
+    symbol_exprs = [(s,v) for s,v in range_constraints.items() if not s.is_symbol]
+    for sym, constraint in symbol_variables:
+      symbolic_shape = _build_symbolic_shape(sym, constraint, symbolic_shapes)
+      symbolic_shapes[str(sym)] = symbolic_shape
+    for sym, constraint in symbol_exprs:
+      symbolic_shape = _build_symbolic_shape(sym, constraint, symbolic_shapes)
+      symbolic_shapes[str(sym)] = symbolic_shape
```

Comment:
Can the for loops be merged? The bodies seem to be the same.

Reply:
Yeah just need to enforce that symbol_variable comes first:

```
    symbol_variables = [(s,v) for s,v in range_constraints.items() if s.is_symbol]
    symbol_exprs = [(s,v) for s,v in range_constraints.items() if not s.is_symbol]
    for sym, constraint in symbol_variables + symbol_exprs:
```

Alt would be using sort with comparator, but I'm not sure if there's a good way to sort on booleans (+not sure I like implying an ordering of bools..), so perhaps this is the best

Reply:
Ah, I see. I think sorting here would be a good fit actually because we want to enforce an ordering and the ordering for booleans is `False < True`. I like what you did with `symbol_variables + symbol_exprs` though.